### ISSUE TITLE
Memory leak via traces

### ISSUE THREAD

**Issue by barrucadu** [barrucadu/dejafu#90]
See: http://svgur.com/s/2VW

Given the linear growth of things in `Test.DejaFu.Conc.*`, this is probably in `runThreads`.

**Comment by barrucadu** [comment]
It turns out that traces are a *far* worse memory leak than I thought.  Here's a very simple benchmark:

```haskell
module Main where

import Control.Concurrent.Classy
import Data.Foldable (for_)
import qualified Test.DejaFu as D
import qualified Test.DejaFu.SCT as D
import qualified System.Random as R

main :: IO ()
main = do
  let gen = R.mkStdGen 42

  putStr "Contended MVar: "
  results <- D.runSCT (D.randomly gen 100) D.defaultMemType contendedMVar
  print (length results)

-- | 1 producer and 20 consumers, with the main thread killed when the
-- final value is consumed.
contendedMVar :: MonadConc m => m ()
contendedMVar = do
  threadId <- myThreadId
  mvar     <- newEmptyMVar

  let maxval = 150
  let go     = takeMVar mvar >>= \x -> if x == maxval then killThread threadId else go

  for_ [1..20] . const $ fork go
  fork $ for_ [1..maxval] (putMVar mvar)

  takeMVar =<< newEmptyMVar
```

And here are the heap profiles with and without traces: [with-traces.pdf](https://github.com/barrucadu/dejafu/files/1205634/with-traces.pdf) [without-traces.pdf](https://github.com/barrucadu/dejafu/files/1205635/without-traces.pdf).  Traces were disabled in the obvious way:

```diff
diff --git a/dejafu/Test/DejaFu/SCT.hs b/dejafu/Test/DejaFu/SCT.hs
index 072522b..4e5e08d 100755
--- a/dejafu/Test/DejaFu/SCT.hs
+++ b/dejafu/Test/DejaFu/SCT.hs
@@ -476,11 +476,11 @@ sctWeightedRandom memtype g0 lim0 use0 conc = go g0 (max 0 lim0) (max 1 use0) M.
   go _ 0 _ _ = pure []
   go g n 0 _ = go g n (max 1 use0) M.empty
   go g n use ws = do
-    (res, s, trace) <- runConcurrent (randSched $ randomR (1, 50))
-                                     memtype
-                                     (initialRandSchedState (Just ws) g)
-                                     conc
-    ((res, trace):) <$> go (schedGen s) (n-1) (use-1) (schedWeights s)
+    (res, s, _) <- runConcurrent (randSched $ randomR (1, 50))
+                                 memtype
+                                 (initialRandSchedState (Just ws) g)
+                                 conc
+    ((res, []):) <$> go (schedGen s) (n-1) (use-1) (schedWeights s)
```

This is pretty bad.  Quite possibly something can be done to improve the memory usage of the average trace, but one possibility for a huge win is to push the predicate inside the SCT function, and only keep failing results rather than all results.  Something like this:

```diff
diff --git a/dejafu/Test/DejaFu/SCT.hs b/dejafu/Test/DejaFu/SCT.hs
index 072522b..8bfb11b 100755
--- a/dejafu/Test/DejaFu/SCT.hs
+++ b/dejafu/Test/DejaFu/SCT.hs
@@ -469,10 +469,12 @@ sctWeightedRandom :: (MonadRef r n, RandomGen g)
   -- ^ The number of executions to perform.
   -> Int
   -- ^ The number of executions to use the same set of weights for.
+  -> (Either Failure a -> Bool)
+  -- ^ The property of interest.  Only results failing this are returned.
   -> ConcT r n a
   -- ^ The computation to run many times.
   -> n [(Either Failure a, Trace)]
-sctWeightedRandom memtype g0 lim0 use0 conc = go g0 (max 0 lim0) (max 1 use0) M.empty where
+sctWeightedRandom memtype g0 lim0 use0 p conc = go g0 (max 0 lim0) (max 1 use0) M.empty where
   go _ 0 _ _ = pure []
   go g n 0 _ = go g n (max 1 use0) M.empty
   go g n use ws = do
@@ -480,7 +482,8 @@ sctWeightedRandom memtype g0 lim0 use0 conc = go g0 (max 0 lim0) (max 1 use0) M.
                                      memtype
                                      (initialRandSchedState (Just ws) g)
                                      conc
-    ((res, trace):) <$> go (schedGen s) (n-1) (use-1) (schedWeights s)
+    let rest = go (schedGen s) (n-1) (use-1) (schedWeights s)
+    if p res then rest else ((res, trace):) <$> rest
```

(profile with the predicate `const True`: [predicate-bool.pdf](https://github.com/barrucadu/dejafu/files/1205661/predicate-bool.pdf))

Though that wouldn't help if you wanted, say, all the results.  Another design is to have a predicate like this:

```diff
diff --git a/dejafu/Test/DejaFu/SCT.hs b/dejafu/Test/DejaFu/SCT.hs
index 072522b..9e3a674 100755
--- a/dejafu/Test/DejaFu/SCT.hs
+++ b/dejafu/Test/DejaFu/SCT.hs
@@ -452,6 +452,8 @@ sctUniformRandom memtype g0 lim0 conc = go g0 (max 0 lim0) where
                                      conc
     ((res, trace):) <$> go (schedGen s) (n-1)

+data How = Discard | KeepResult | KeepResultAndTrace
+
 -- | SCT via weighted random scheduling.
 --
 -- Schedules are generated by assigning to each new thread a random
@@ -469,10 +471,12 @@ sctWeightedRandom :: (MonadRef r n, RandomGen g)
   -- ^ The number of executions to perform.
   -> Int
   -- ^ The number of executions to use the same set of weights for.
+  -> (Either Failure a -> How)
+  -- ^ Selectively exclude results and traces from the final result list.
   -> ConcT r n a
   -- ^ The computation to run many times.
   -> n [(Either Failure a, Trace)]                                                                                                                                                                        -sctWeightedRandom memtype g0 lim0 use0 conc = go g0 (max 0 lim0) (max 1 use0) M.empty where
+sctWeightedRandom memtype g0 lim0 use0 p conc = go g0 (max 0 lim0) (max 1 use0) M.empty where
   go _ 0 _ _ = pure []
   go g n 0 _ = go g n (max 1 use0) M.empty
   go g n use ws = do                                                                                                                                                                                      @@ -480,7 +484,11 @@ sctWeightedRandom memtype g0 lim0 use0 conc = go g0 (max 0 lim0) (max 1 use0) M.
                                      memtype
                                      (initialRandSchedState (Just ws) g)
                                      conc                                                                                                                                                                 -    ((res, trace):) <$> go (schedGen s) (n-1) (use-1) (schedWeights s)                                                                                                                                    +    let rest = go (schedGen s) (n-1) (use-1) (schedWeights s)                                                                                                                                             +    case p res of                                                                                                                                                                                         +      Discard -> rest                                                                                                                                                                                     +      KeepResult -> ((res, []):) <$> rest
+      KeepResultAndTrace -> ((res, trace):) <$> rest
```

(profile with the predicate `const KeepResult`: [predicate-how.pdf](https://github.com/barrucadu/dejafu/files/1205673/predicate-how.pdf))

This option would make it possible to have a far more efficient `resultsSet` function.

Whatever extra behaviour is decided on, it can also be added to `sctUniformRandom` and `sctBound`, though the latter probably wouldn't see much benefit until #89 or #64 is implemented.

**Comment by barrucadu** [comment]
Note:  there is definitely a memory leak here (in addition to traces eating a lot of space), as `Bin` values (from Data.Map) shouldn't be escaping.

**Comment by barrucadu** [comment]
5cb8b4e170c6725d74b7e3efc5b125666c0bd11b doesn't seem to solve the original problem:  
![bad](https://user-images.githubusercontent.com/75235/29118003-46d8f42e-7cf8-11e7-8509-fbdce76fc351.png) whereas just disabling traces (as in the first diff) does: 
![good](https://user-images.githubusercontent.com/75235/29118017-52f8eda4-7cf8-11e7-819d-ab32e6ef64da.png)

The tests are being run through `hunit-dejafu`, maybe there's some laziness being introduced which scuppers things.

**Comment by barrucadu** [comment]
Original issue closed by merging `discard` in 50712c54cd9fb9f4a308893645ec71ba0897b455.
